extensions/mapstore/src/main/java/com/hazelcast/mapstore/GenericMapStore.java

```diff
@@ -0,0 +1,518 @@
+            SqlColumnMetadata column = metadata.getColumn(i);
+
+            if (column.getName().equals(properties.idColumn) && !properties.idColumnInColumns) {
+                continue;
+            }
+
+            switch (column.getType()) {
+                case VARCHAR:
+                    builder.setString(column.getName(), row.getObject(i));
+                    break;
+
+                case INTEGER:
+                    builder.setInt32(column.getName(), row.getObject(i));
+                    break;
+
+                default:
+                    throw new HazelcastException("Column type " + column.getType() + " not supported");
+            }
+        }
+
+        return builder.build();
+    }
+
+    /**
+     * Size of the {@code keys} collection is limited by {@link ClusterProperty#MAP_LOAD_CHUNK_SIZE}
+     */
+    @Override
+    public Map<K, GenericRecord> loadAll(Collection<K> keys) {
+        awaitInitFinished();
+
+        Object[] keysArray = keys.toArray();
+
+        try (SqlResult queryResult = sql.execute(queries.loadAll(keys.size()), keysArray)) {
+            Iterator<SqlRow> it = queryResult.iterator();
+
+            Map<K, GenericRecord> result = new HashMap<>();
+            while (it.hasNext()) {
+                SqlRow row = it.next();
+                K id = row.getObject(properties.idColumn);
+                GenericRecord record = convertRowToGenericRecord(id, row);
+                result.put(id, record);
+            }
+            return result;
+        }
+    }
+
+    @Override
+    public Iterable<K> loadAllKeys() {
+        awaitInitFinished();
+
+        SqlResult keysResult = sql.execute(queries.loadAllKeys());
+
+        return () -> new MappingClosingIterator<>(
+                keysResult.iterator(),
+                (SqlRow row) -> row.getObject(properties.idColumn),
+                keysResult::close
+        );
+    }
+
+    @Override
+    public void store(K key, GenericRecord value) {
+        awaitInitFinished();
+
+        int idPos = -1;
+        Object[] params = new Object[columnMetadata.size()];
+        for (int i = 0; i < columnMetadata.size(); i++) {
+            SqlColumnMetadata cm = columnMetadata.get(i);
+            if (cm.getName().equals(properties.idColumn)) {
+                idPos = i;
+            }
+            switch (cm.getType()) {
+                case VARCHAR:
+                    params[i] = value.getString(cm.getName());
+                    break;
+
+                case INTEGER:
+                    params[i] = value.getInt32(cm.getName());
+                    break;
+
+                default:
+                    throw new HazelcastException("Column type " + cm.getType() + " not supported");
+            }
+        }
+        try (SqlResult ignored = sql.execute(queries.storeInsert(), params)) {
+        } catch (Exception e) {
+            if (e.getMessage().contains("Unique index or primary key violation")) {
```

Comment:
Could we use SQL features like `INSERT ... ON DUPLICATE KEY UPDATE` or `INSERT ... ON CONFLICT`

Reply:
I wanted to, but
- this is not well standardized - mysql and postgres have different syntax
- it is not supported by our SQL engine